benchopt/utils/safe_import.py

```diff
@@ -69,43 +132,29 @@ def __init__(self):
         self._benchmark_dir = BENCHMARK_DIR
 
     def __enter__(self):
-        # Skip context if necessary to speed up import
-        if SKIP_IMPORT:
-            # See https://stackoverflow.com/questions/12594148/skipping-execution-of-with-block  # noqa
-            sys.settrace(lambda *args, **keys: None)
-            frame = sys._getframe(1)
-            frame.f_trace = self.trace
-            return self
+        # Mock import to speed up import
+        if MOCK_ALL_IMPORT or MOCK_FAILED_IMPORT:
+            sys.meta_path.insert(0, MockFinder())
 
         # Catch the import warning except if install errors are raised.
         if not RAISE_INSTALL_ERROR:
             self.record.__enter__()
-        return self
 
-    def trace(self, frame, event, arg):
-        raise SkipWithBlock()
+        return self
 
     def __exit__(self, exc_type, exc_value, tb):
-
-        if SKIP_IMPORT:
+        if MOCK_ALL_IMPORT or MOCK_FAILED_IMPORT:
```

Comment:
This should only be if `MOCK_ALL_IMPORT` or if at least one import has been mocked.